README
EXTRACT REASONABLE TIME RECORDS FROM A GIT REPOSITORY

This tool goes through a GIT repository's commit log and prints a CSV dump of per developer, per day working time based on a few assumptions:

    - A series of commits within a 3 hour window are part of the same development session
    - A single commit (or the first commit of the session) is considered to represent 30 minutes of work time
    - The more frequent a developer commits to the repository while working, the more accurate the time report will be
    
This script is based on previous code published publicly by Sharad at http://www.tatvartha.com/2010/01/generating-time-entry-from-git-log/. However, it has been adapted to run without Rails from the command line. The portions of the code written by Rietta are licensed under the terms of the BSD license (see section 3 below).
    
-----------------------------------------
1. Prerequisites
-----------------------------------------

This script is designed to work in Ruby 1.9 from the command line. It does not require Rails, but rather is design to work with pure Ruby and a couple of support gems, namely:

    - git
    - ostruct
    - logger
    - csv

Most of the requirements should be present on your base Ruby install.  However, you will need to first install a git gem.

For example, run the following:

    gem install git
    
-----------------------------------------
2. Running
-----------------------------------------

    cd /path/to/your/repository
    time_extract.rb > time_log.csv
    

-----------------------------------------
3. LICENSE TERMS (BSD License)
-----------------------------------------

Copyright (c) 2012 Rietta Inc. All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.  
    


History.txt
=== 0.3.2 / 2015-05-20
  Multiple small changes, compact and total-project summary options added.

=== 0.3.0 / 2014-11-13
  Major update. Command line configurable assumptions and multiple author support.
=== 0.2.3
=== 0.1.0 / 2012-05-05

* 1 major enhancement

  * Birthday of this git_time_extractor gem.

=== 0.1.0 / 2012-02-15
* Initial code release on https://github.com/rietta/git_time_extractor/ - not a gem.

test/test_git_time_extractor.rb
require "test/unit"
require "git_time_extractor"

# TODO: Test cases are needed.  It is not immediately clear how to best build a test suite.
# Supposedly if a GIT repository was distributed within this Gem, it could be used to verify
# that this gem can write out a log.
#
class TestGitTimeExtractor < Test::Unit::TestCase
  def test_sanity
    assert_equal "hi", GitTimeExtractor.say_hi  
  end
end


lib/author.rb
class Author

  attr_accessor :commits, :worklog, :rows, :options

  def initialize(options)
    @options  = options
    @commits  = Array.new
    @worklog  = Hash.new
    @rows     = Array.new
  end

  # Each commit is added into the list
  def add_commit(commit)
    @commits << commit
  end

  def total_commits
    return @commits.length
  end

  def total_working_minutes
    # Go through the work log
    total = 0
    @worklog.keys.sort.each do |date|
      if date_in_filter_range(date)
        total += @worklog[date].duration.to_i
      end
    end # worklog each
    return total
  end
  
  def total_working_hours
    return (total_working_minutes() / 60.0).round(1)
  end

  # Then the tabulation is called after the commits are added into this author's list
  def tabulate_days
    @commits.each_with_index do |commit, index|
       roll_up_to_days commit, index
    end

    # Go through the work log
    @worklog.keys.sort.each do |date|
      if date_in_filter_range(date)
        @rows << summarize(@worklog[date], date)
      end
    end # worklog each
  end # tabulate_days

  def summarize(summary, date)
    start_time = DateTime.parse(date.to_s)
    duration_in_seconds = (summary.duration.to_f * 60.0).round(0)
    duration_in_minutes = summary.duration.to_i
    duration_in_hours   = (summary.duration / 60.0).round(1)

    return summarize_helper(
                              start_time,
                              duration_in_seconds,
                              duration_in_minutes,
                              duration_in_hours,
                              summary
                           )
  end # summarize

  def summarize_helper(
                        start_time,
                        duration_in_seconds,
                        duration_in_minutes,
                        duration_in_hours,
                        summary
                      )
    [
      start_time.strftime("%m/%d/%Y"),
      summary.commit_count,
      summary.pivotal_stories.count,
      duration_in_minutes,
      duration_in_hours,
      summary.author.name,
      summary.author.email,
      project_name,
      summary.message,
      summary.pivotal_stories.map(&:inspect).join('; '),
      start_time.strftime("%W").to_i,
      start_time.strftime("%Y").to_i
    ]
  end # summarize_helper

  def roll_up_to_days(commit, index)
    # Get the appropriate worklog
    author_date = commit.author_date.to_date

    #puts "DEBUG: Reading #{commit} #{author_date} from #{commit.author.email}"
    daylog = get_or_create_new_daylog(author_date)
    @worklog[author_date] = process_day_log(commit, index, daylog)
  end # roll_up_to_days

  def get_or_create_new_daylog(author_date)
    if @worklog[author_date]
      return @worklog[author_date]
    else
      return OpenStruct.new(
                      :date => author_date,
                      :duration => 0,
                      :commit_count => 0,
                      :pivotal_stories => Set.new
                    )
    end # if
  end # daylog

  def pivotal_ids(text)
    ::PivotalIdsExtractor.new(text).stories
  end

  def process_day_log(commit, index, daylog)
    #puts "DEBUG: Processing #{commit} - #{commit.author}"
    daylog.author = commit.author
    daylog.message = "#{daylog.message} --- #{commit.message}"
    daylog.duration = daylog.duration + calc_duration_in_minutes(@commits, index)

    # The git commit count
    daylog.commit_count += 1

    # Pivotal Stories
    stories = pivotal_ids(commit.message)
    if stories
      # It's a set, so each story only gets added once per day
      stories.each do |sid|
        daylog.pivotal_stories << sid
      end # each
    end # if stories

    return daylog
  end # process_day_log

  # Calculate the duration of work in minutes
  def calc_duration_in_minutes(log_entries, index)
    if index > 1
      # Compute the time between this and this previous entry
      return compute_commit_time_duration(log_entries[index], log_entries[index - 1])
    else
      # This is the first commit in the log
      return @options[:initial_effort_mins].to_f
    end
    return duration
  end # calc_duration_in_minutes

  def compute_commit_time_duration(commit, previous_commit)
    # Default duration in Ruby is in seconds, so lets make it minutes
    duration = (commit.author_date - previous_commit.author_date) / 60.0

    #initial_effort_mins: 30, session_duration_hrs: 3, max_commits: 1000
    if duration > @options[:session_duration_hrs].to_f * 60.0
      # first commit in this session
      duration = @options[:initial_effort_mins].to_f
    elsif duration < 0
      # probably a merge.
      duration = @options[:merge_effort_mins].to_f
    end

    return duration.to_f
  end # compute_commit_time_duration

  def project_name
    @options[:project]
  end

  def date_in_filter_range(date_to_check)
    if @options[:filter_by_year] and date_to_check
      date_to_check.strftime("%Y").to_i == @options[:filter_by_year]
    else
      true
    end
  end

end # class

lib/git_time_extractor.rb
#
# Extract Reasonable Developer Time Records from a GIT Repository's Commit Log
#
# This is inspired by a RAKE task publicly posted by Sharad at
# http://www.tatvartha.com/2010/01/generating-time-entry-from-git-log/.
# However, it has been adapted to run without Rails from the command line.
#
# Portions (C) 2014 Rietta Inc. and licensed under the terms of the BSD license.
#
class GitTimeExtractor
  VERSION = '0.3.2'

  require 'autoload'
  require 'set'

  attr_accessor :options, :summary, :authors

  def initialize( opts = {
                          project: "Untitled",
                          path_to_repo: "./",
                          output_file: "-",
                          initial_effort_mins: 30,
                          merge_effort_mins: 30,
                          session_duration_hrs: 3,
                          max_commits: 1000
                        } )
    @options = opts
    @authors_hash = Hash.new
    @authors      = Array.new
  end

  def path_to_git_repo
    options[:path_to_repo]
  end

  def path_to_output_file
    options[:output_file]
  end

  def project_name
    options[:project]
  end

  def load_git_log_entries(path_to_git_repo)
    # Open the GIT Repository for Reading
    logger = Logger.new(STDOUT)
    logger.level = Logger::WARN
    g = Git.open(path_to_git_repo, :log => logger)
    logs = g.log(@options[:max_commits])
    return logs.entries.reverse
  end

  # Scan through the commit log entries and split into each author's accounting
  def distribute_entries_to_authors(log_entries)
    # Go through the GIT commit records and construct the time
    log_entries.each_with_index do |commit, index|
      key = commit.author.email
       if @authors_hash[key].nil?
         @authors_hash[key] = Author.new(@options)
       end
       @authors_hash[key].add_commit(commit)
    end # log_entries.each_with_index

    if @authors_hash
      @authors_hash.each do |key, value|
        if value
          value.tabulate_days
          @authors << value
          #puts "DEBUG: Looking at #{key}, #{value}, with #{value.rows.count} commits"
        end
      end
    end

    #puts "DEBUG: #{@authors.count} authors"

    return @authors
  end # distribute_entries_to_authors

  def prepare_rows_for_csv
    rows = Array.new
    rows << header_row_template()
    @authors.each do |author|
      author.rows.each do |author_row|
        rows << author_row
      end
    end
    return rows
  end

  #
  # Go through the GIT commit log, to compute the elapsed working time of each committing developer, based
  # on a few assumptions:
  #
  # (1) A series of commits within a 3 hour window are part of the same development session
  # (2) A single commit (or the first commit of the session) is considered to represent 30 minutes of work time
  # (3) The more frequent a developer commits to the repository while working, the more accurate the time report will be
  #
  #
  def process_git_log_into_time
    distribute_entries_to_authors( load_git_log_entries( path_to_git_repo ) )
    return prepare_rows_for_csv
  end # process_git_log_into_time


  def prepare_rows_for_summary_csv
    rows = Array.new
    rows << summary_header_row_template()
    commits = 0
    hours = 0
    authors = Array.new
    @authors.each do |author|
      commits += author.total_commits()
      hours += author.total_working_hours()
      authors << author.commits[0].author.name + " (" + author.commits[0].author.email + ")"
    end
    rows << [
      hours,
      commits,
      authors.join(";")
    ]
    return rows
  end

  #
  # create a summary of the computed times
  #
  def process_git_log_into_summary
    distribute_entries_to_authors( load_git_log_entries( path_to_git_repo ) )
    return prepare_rows_for_summary_csv
  end # process_git_log_into_time


  def prepare_rows_for_author_summary_csv
    rows = Array.new
    rows << author_summary_header_row_template()
    commits = 0
    hours = 0
    authors = Array.new
    @authors.each do |author|
      rows << [
        author.total_commits(),
        author.total_working_hours(),
        author.commits[0].author.name,
        author.commits[0].author.email,
      ]
    end
    return rows
  end

  #
  # create a summary per author
  #
  def process_git_log_into_author_summary
    distribute_entries_to_authors( load_git_log_entries( path_to_git_repo ) )
    return prepare_rows_for_author_summary_csv
  end # process_git_log_into_time

  def process
    if options[:project_total]
      process_git_log_into_summary
    elsif options[:compact]
      process_git_log_into_author_summary
    else
      process_git_log_into_time
    end
  end

  #####################################
  private

  def header_row_template
    [
      'Date',
      'Git Commits Count',
      'Pivotal Stories Count',
      'Minutes',
      'Hours',
      'Person',
      'Email',
      'Project',
      'Notes',
      'Pivotal Stories',
      'Week Number',
      'Year'
    ]
  end # header_row_template

  def author_summary_header_row_template
    [
      'Total Git Commits Count',
      'Total Hours',
      'Person',
      'Email',
      # 'From Date',
      # 'To Date',
    ]
  end # summary_header_row_template

  def summary_header_row_template
    [
      # 'From Date',
      # 'To Date',
      'Total Git Commits Count',
      'Total Hours',
      'Collaborators',
    ]
  end # summary_header_row_template

end # class GitTimeExtractor


lib/pivotal_ids_extractor.rb
class PivotalIdsExtractor

  attr_accessor :source_text, :stories

  #  --- [#62749778] New Email Page --- Merge branch 'development' of bitbucket.org:rietta/roofregistry-web into development --- [#62749778] Roughed out email form. --- Added delete Attachment functionality --- Merge branch 'development' of bitbucket.org:rietta/roofregistry-web into development --- [#62749778] Refactored controller to be plural. --- [#62749778] Added to the Email model. --- [62749778] The email this report view formatting. --- [#62749778] Breadcrumbs in the navigation. --- [#62749778] The Emails controller routes. --- The report list is now sorted with newest first - and it shows how long ago that the change was made. --- [#62749778] The share link is bold. --- [#62749778] Recipient parsing and form fields --- [#62749778] List of emails that have received it. --- [#62749778] The email form will validate that at least one email is provided. --- [#62749778] Send Roof Report AJAX form. --- [#62749778] Default messages and the mailer --- [Finishes #62749778] The emails are sent! --- removed delete from show --- added txt and xpf to permitted file types --- Attachments can only be deleted by the owner of the roof report. --- Merge branch 'development' of bitbucket.org:rietta/roofregistry-web into development --- The test server is using production. --- Returns all recommended options across all sections with roof_report.recommedations --- patial commit --- Finished summary section --- Added caps to permitted --- added to_s to inspection --- E-mail spec is not focused at the moment. --- Merge branch 'development' of bitbucket.org:rietta/roofregistry-web into development --- fixed a few bugs --- Merge branch 'development' of bitbucket.org:rietta/roofregistry-web into development --- Disable ajax save. --- Merge branch 'development' of bitbucket.org:rietta/roofregistry-web into development
  # s = "[#62749778] [#62749778] [#6274977] [#1] [#231]"
  # m = s.scan /\[[A-Za-z ]{0,20}#[0-9]{1,20}\]/

  def initialize(text)
    stories = Array.new
    # Extract the unique ids between brackets
    if text
      # Require brackets
      #candidates = text.scan /\[[A-Za-z \t]{0,20}#[0-9]{1,35}[ \t]{0,5}\]/
      candidates = text.scan /[A-Za-z \t]{0,20}#[0-9]{1,35}[ \t]{0,5}/
      if candidates
        candidates.uniq.each do |story|
          story_num = story.match(/[0-9]{1,35}/).to_s.to_i
          stories << story_num if story_num > 0
        end
      end
    end
    @source_text  = text
    @stories      = stories.sort
  end


end

time_extract.rb
#!/usr/bin/env ruby
#
# Extract Reasonable Developer Time Records from a GIT Repository's Commit Log
#
# This is inspired by a RAKE task publicly posted by Sharad at 
# http://www.tatvartha.com/2010/01/generating-time-entry-from-git-log/. 
# However, it has been adapted to run without Rails from the command line.
#
# This script is designed to work from the command line with Ruby 1.9.3. It requires
# the git gem.  Install it on your system with:
#   gem install git
#
# Portions (C) 2012 Rietta Inc. and licensed under the terms of the BSD license.
#

require 'rubygems'
require 'ostruct'
require 'logger'
require 'git'
require 'csv'

def process_git_log_into_time(path_to_git_repo = "./", path_to_output_file = "-", project_name = "")
  
  raise "Output path not yet implemented." if "-" != path_to_output_file
  
  # Open the GIT Repository for Reading
  logger = Logger.new(STDOUT)
  logger.level = Logger::WARN
  g = Git.open(path_to_git_repo, :log => logger)
  logs = g.log(1000)
  log_entries = logs.entries.reverse
  worklog = {}

  # Go through the GIT commit records and construct the time
  log_entries.each_with_index do |commit, index|
    author_date = commit.author_date.to_date
    daylog = worklog[author_date] || OpenStruct.new(:date => author_date, :duration => 0)
    daylog.author = commit.author
    daylog.message = "#{daylog.message} --- #{commit.message}"
    daylog.duration = daylog.duration + calc_duration_in_minutes(log_entries, index)
    worklog[author_date] = daylog
  end # log_entries
  
  # Print the header row for the CSV
  puts [
      'Date',
      'Minutes',
      'Hours',
      'Person',
      'Email',
      'Project',
      'Notes',
      'Week Number',
      'Year'
      ].to_csv
  
  # Go through the work log  
  worklog.keys.sort.each do |date|
      
      start_time = DateTime.parse(date.to_s)
      duration_in_seconds = (worklog[date].duration.to_f * 60.0).round(0)
      duration_in_minutes = worklog[date].duration.to_i
      duration_in_hours = (worklog[date].duration / 60.0).round(1)
      
      stop_time = start_time + duration_in_seconds
      row = [
            start_time.strftime("%m/%d/%Y"), 
            duration_in_minutes,
            duration_in_hours,
            worklog[date].author.name,
            worklog[date].author.email,
            project_name,
            worklog[date].message,
            start_time.strftime("%W").to_i,
            start_time.strftime("%Y").to_i]
      puts row.to_csv
  end # worklog each

end # process_git_log_into_time
 
def calc_duration_in_minutes(log_entries, index)
  commit = log_entries[index]
  if index > 1
    previous_commit = log_entries[index-1]
    
    # Default duration in Ruby is in seconds
    duration = commit.author_date - previous_commit.author_date
    # ASSUMPTION: if the gap between 2 commits is more than 3 hours, reduce it to 1/2 hour
    duration = 30 * 60 if duration > 3 * 3600
  else
    # ASSUMPTION: first commit took 1/2 hour
    duration = 30 * 60
  end
  
  return duration.to_f / 60.0
end # calc_duration


####################################
## Command Line Interface
####################################

#puts "\n# Arguments #{ARGV.length}\n"
if ARGV.empty?
  output_file = "-"
  path_to_repo = Dir.pwd
  project_name = ""
elsif ARGV.length == 2
  output_file = ARGV.pop
  path_to_repo = ARGV.pop
elsif ARGV.length == 1
  output_file = "-"
  path_to_repo = ARGV.pop
else
  puts "Usage: time_extract PROJECT_NAME [PATH_TO_REPO] [OUTPUT_FILE]"
  exit 0
end

#path_to_repo = File.expand_path(File.dirname(path_to_repo)) if nil != path_to_repo && "" != path_to_repo
#puts "\n\nGit Repo Path: #{path_to_repo}\nOutput: #{output_file}\n"
process_git_log_into_time(path_to_repo, output_file)

